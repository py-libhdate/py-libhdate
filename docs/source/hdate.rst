hdate package
=============

Submodules
----------

hdate.common module
-------------------

.. automodule:: hdate.common
   :members:
   :undoc-members:
   :show-inheritance:

hdate.converters module
-----------------------

.. automodule:: hdate.converters
   :members:
   :undoc-members:
   :show-inheritance:

hdate.date module
-----------------

.. automodule:: hdate.date
   :members:
   :undoc-members:
   :show-inheritance:

hdate.htables module
--------------------

.. automodule:: hdate.htables
   :members:
   :undoc-members:
   :show-inheritance:

hdate.zmanim module
-------------------

.. automodule:: hdate.zmanim
   :members:
   :undoc-members:
   :show-inheritance:

Module contents
---------------

.. automodule:: hdate
   :members:
   :undoc-members:
   :show-inheritance:
