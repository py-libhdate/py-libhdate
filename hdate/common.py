"""Small helper classes."""

from dataclasses import dataclass
from datetime import tzinfo

import pytz

from hdate.htables import Months


class BaseClass:
    """Implement basic functionality for all classes."""

    def __eq__(self, other):
        """Override equality operator."""
        if isinstance(other, self.__class__):
            return self.__dict__ == other.__dict__
        return False

    def __ne__(self, other):
        """Override inequality operator."""
        return not self.__eq__(other)


@dataclass
class HebrewDate(BaseClass):
    """Define a Hebrew date object."""

    year: int
    month: Months
    day: int

    def __post_init__(self):
        self.month = (
            self.month if isinstance(self.month, Months) else Months(self.month)
        )


class Location(BaseClass):
    """Define a geolocation for Zmanim calculations."""

    # pylint: disable=too-many-arguments
    def __init__(
        self,
        name="Jerusalem",
        latitude=31.778,
        longitude=35.235,
        timezone="Asia/Jerusalem",
        altitude=754,
        diaspora=False,
    ):
        """Initialitze the location object."""
        self._timezone = None
        self.name = name
        self.latitude = latitude
        self.longitude = longitude
        self.timezone = timezone
        self.altitude = altitude
        self.diaspora = diaspora

    def __repr__(self):
        """Return a representation of Location for programmatic use."""
        return (
            f"Location(name='{self.name}', latitude={self.latitude}, "
            f"longitude={self.longitude}, timezone='{self.timezone}', "
            f"altitude={self.altitude}, diaspora={self.diaspora})"
        )

    @property
    def timezone(self):
        """Return the timezone."""
        return self._timezone

    @timezone.setter
    def timezone(self, value):
        """Set the timezone."""
        self._timezone = value if isinstance(value, tzinfo) else pytz.timezone(value)
