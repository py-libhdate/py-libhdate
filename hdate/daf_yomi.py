"""Daf Yomi module."""

import datetime as dt
from dataclasses import dataclass
from typing import ClassVar

from hdate.gematria import hebrew_number
from hdate.translator import TranslatorMixin


@dataclass(frozen=True)
class Masechta(TranslatorMixin):
    """Masechta object."""

    name: str
    pages: int

    def __str__(self) -> str:
        name = self.get_translation(self.name)
        daf = hebrew_number(self.pages + 2, language=self._language, short=True)
        return f"{name} {daf}"


@dataclass
class DafYomiDatabase:
    """Database of Masechtos."""

    # The first few cycles were only 2702 blatt. After that it became 2711. Even with
    # that, the math doesn't play nicely with the dates before the 11th cycle :(
    # From cycle 11 onwards, it was simple and sequential
    _start_date: ClassVar[dt.date] = dt.date(1997, 9, 29)
    _masechtot: ClassVar[list[Masechta]]

    @classmethod
    def register(cls, masechtot: list[Masechta]) -> None:
        """Regisetr a list of masechtos."""
        cls._masechtot = masechtot

    @classmethod
    def cycle_length(cls) -> int:
        """Return the length of a full cycle."""
        return sum(masechta.pages for masechta in cls._masechtot)

    def lookup(self, date: dt.date) -> Masechta:
        """Return the start date of a full cycle."""
        days_since_start = (date - self._start_date).days
        page_number = days_since_start % self.cycle_length()

        masechta_index = 0
        for masechta in self._masechtot:
            if page_number < masechta.pages:
                break
            page_number -= masechta.pages
            masechta_index += 1

        return Masechta(self._masechtot[masechta_index].name, page_number)


DAF_YOMI_MESECHTOS = (
    Masechta("berachos", 63),
    Masechta("shabbos", 156),
    Masechta("eruvin", 104),
    Masechta("pesachim", 120),
    Masechta("shekalim", 21),
    Masechta("yoma", 87),
    Masechta("succah", 55),
    Masechta("beitzah", 39),
    Masechta("rosh_hashanah", 34),
    Masechta("taanis", 30),
    Masechta("megillah", 31),
    Masechta("moed_katan", 28),
    Masechta("chagigah", 26),
    Masechta("yevamos", 121),
    Masechta("kesubos", 111),
    Masechta("nedarim", 90),
    Masechta("nazir", 65),
    Masechta("sotah", 48),
    Masechta("gittin", 89),
    Masechta("kiddushin", 81),
    Masechta("bava_kamma", 118),
    Masechta("bava_metzia", 118),
    Masechta("bava_basra", 175),
    Masechta("sanhedrin", 112),
    Masechta("makkos", 23),
    Masechta("shevuos", 48),
    Masechta("avodah_zarah", 75),
    Masechta("horayos", 13),
    Masechta("zevachim", 119),
    Masechta("menachos", 109),
    Masechta("chullin", 141),
    Masechta("bechoros", 60),
    Masechta("arachin", 33),
    Masechta("temurah", 33),
    Masechta("kereisos", 27),
    Masechta("meilah", 36),
    Masechta("niddah", 72),
)

DafYomiDatabase.register(list(DAF_YOMI_MESECHTOS))
