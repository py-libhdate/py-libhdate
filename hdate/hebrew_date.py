"""Pure Hebrew date class."""

from dataclasses import dataclass

from hdate.htables import Months


@dataclass
class HebrewDate:
    """Define a Hebrew date object."""

    year: int
    month: Months
    day: int

    def __post_init__(self):
        self.month = (
            self.month if isinstance(self.month, Months) else Months(self.month)
        )
