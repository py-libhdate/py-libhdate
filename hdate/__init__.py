"""
Jewish calendrical date and times for a given location.

HDate calculates and generates a representation either in English or Hebrew
of the Jewish calendrical date and times for a given location
"""
from hdate.common import HebrewDate, Location
from hdate.date import HDate
from hdate.htables import HolidayTypes
from hdate.zmanim import Zmanim

__all__ = ["HDate", "Zmanim", "HebrewDate", "Location", "HolidayTypes"]
