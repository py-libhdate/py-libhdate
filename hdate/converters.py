"""Methods for going back and forth between various calendars."""
import datetime

from hdate.common import HebrewDate
from hdate.htables import Months


def get_chalakim(hours, parts):
    """Return the number of total parts (chalakim)."""
    return (hours * PARTS_IN_HOUR) + parts


PARTS_IN_HOUR = 1080
PARTS_IN_DAY = 24 * PARTS_IN_HOUR
PARTS_IN_WEEK = 7 * PARTS_IN_DAY
PARTS_IN_MONTH = PARTS_IN_DAY + get_chalakim(12, 793)  # Fix for regular month


def _days_from_3744(hebrew_year):
    """Return: Number of days since 3,1,3744."""
    # Start point for calculation is Molad new year 3744 (16BC)
    years_from_3744 = hebrew_year - 3744
    molad_3744 = get_chalakim(1 + 6, 779)  # Molad 3744 + 6 hours in parts

    # Time in months

    # Number of leap months
    leap_months = (years_from_3744 * 7 + 1) // 19
    leap_left = (years_from_3744 * 7 + 1) % 19  # Months left of leap cycle
    months = years_from_3744 * 12 + leap_months  # Total Number of months

    # Time in parts and days
    # Molad This year + Molad 3744 - corrections
    parts = months * PARTS_IN_MONTH + molad_3744
    # 28 days in month + corrections
    days = months * 28 + parts // PARTS_IN_DAY - 2

    # Time left for round date in corrections
    # 28 % 7 = 0 so only corrections counts
    parts_left_in_week = parts % PARTS_IN_WEEK
    parts_left_in_day = parts % PARTS_IN_DAY
    week_day = parts_left_in_week // PARTS_IN_DAY

    # pylint: disable=too-many-boolean-expressions
    # pylint-comment: Splitting the 'if' below might create a bug in case
    # the order is not kept.

    # Molad ד"ר ט"ג
    if (
        (
            leap_left < 12
            and week_day == 3
            and parts_left_in_day >= get_chalakim(9 + 6, 204)
        )
        or
        # Molad ט"פקת ו"טב
        (
            leap_left < 7
            and week_day == 2
            and parts_left_in_day >= get_chalakim(15 + 6, 589)
        )
    ):
        days += 1
        week_day += 1

    # pylint: enable=too-many-boolean-expressions

    # ADU
    if week_day in (1, 4, 6):
        days += 1

    return days


def get_size_of_hebrew_year(hebrew_year):
    """Return: total days in hebrew year."""
    return _days_from_3744(hebrew_year + 1) - _days_from_3744(hebrew_year)


def gdate_to_jdn(date):
    """
    Compute Julian day from Gregorian day, month and year.

    Algorithm from wikipedia's julian_day article.
    Return: The julian day number
    """
    not_jan_or_feb = (14 - date.month) // 12
    year_since_4800bc = date.year + 4800 - not_jan_or_feb
    month_since_4800bc = date.month + 12 * not_jan_or_feb - 3
    jdn = (
        date.day
        + (153 * month_since_4800bc + 2) // 5
        + 365 * year_since_4800bc
        + (year_since_4800bc // 4 - year_since_4800bc // 100 + year_since_4800bc // 400)
        - 32045
    )
    return jdn


def hdate_to_jdn(date):
    """
    Compute Julian day from Hebrew day, month and year.

    Return: julian day number,
            1 of tishrey julians,
            1 of tishrey julians next year
    """
    day = date.day
    month = date.month.value
    if date.month == Months.ADAR_I:
        month = 6
    if date.month == Months.ADAR_II:
        month = 6
        day += 30

    # Calculate days since 1,1,3744
    day = _days_from_3744(date.year) + (59 * (month - 1) + 1) // 2 + day

    # length of year
    length_of_year = get_size_of_hebrew_year(date.year)
    # Special cases for this year
    if length_of_year % 10 > 4 and month > 2:  # long Heshvan
        day += 1
    if length_of_year % 10 < 4 and month > 3:  # short Kislev
        day -= 1
    if length_of_year > 365 and month > 6:  # leap year
        day += 30

    # adjust to julian
    return day + 1715118


def jdn_to_gdate(jdn):
    """
    Convert from the Julian day to the Gregorian day.

    Algorithm from 'Julian and Gregorian Day Numbers' by Peter Meyer.
    Return: day, month, year
    """
    # pylint: disable=invalid-name

    # The algorithm is a verbatim copy from Peter Meyer's article
    # No explanation in the article is given for the variables
    # Hence the exceptions for pylint and for flake8 (E741)

    l = jdn + 68569  # noqa: E741
    n = (4 * l) // 146097
    l = l - (146097 * n + 3) // 4  # noqa: E741
    i = (4000 * (l + 1)) // 1461001  # that's 1,461,001
    l = l - (1461 * i) // 4 + 31  # noqa: E741
    j = (80 * l) // 2447
    day = l - (2447 * j) // 80
    l = j // 11  # noqa: E741
    month = j + 2 - (12 * l)
    year = 100 * (n - 49) + i + l  # that's a lower-case L

    return datetime.date(year, month, day)


def jdn_to_hdate(jdn):
    """Convert from the Julian day to the Hebrew day."""
    # calculate Gregorian date
    date = jdn_to_gdate(jdn)

    # Guess Hebrew year is Gregorian year + 3760
    year = date.year + 3760

    jdn_tishrey1 = hdate_to_jdn(HebrewDate(year, Months.TISHREI, 1))
    jdn_tishrey1_next_year = hdate_to_jdn(HebrewDate(year + 1, Months.TISHREI, 1))

    # Check if computed year was underestimated
    if jdn_tishrey1_next_year <= jdn:
        year = year + 1
        jdn_tishrey1 = jdn_tishrey1_next_year
        jdn_tishrey1_next_year = hdate_to_jdn(HebrewDate(year + 1, Months.TISHREI, 1))

    size_of_year = get_size_of_hebrew_year(year)

    # days into this year, first month 0..29
    days = jdn - jdn_tishrey1

    # last 8 months always have 236 days
    if days >= (size_of_year - 236):  # in last 8 months
        days = days - (size_of_year - 236)
        month = days * 2 // 59
        day = days - (month * 59 + 1) // 2 + 1

        month = month + 4 + 1

        # if leap
        if size_of_year > 355 and month <= 6:
            month = month + 8
    else:  # in 4-5 first months
        # Special cases for this year
        if size_of_year % 10 > 4 and days == 59:  # long Heshvan (day 30)
            month = 1
            day = 30
        elif size_of_year % 10 > 4 and days > 59:  # long Heshvan
            month = (days - 1) * 2 // 59
            day = days - (month * 59 + 1) // 2
        elif size_of_year % 10 < 4 and days > 87:  # short kislev
            month = (days + 1) * 2 // 59
            day = days - (month * 59 + 1) // 2 + 2
        else:  # regular months
            month = days * 2 // 59
            day = days - (month * 59 + 1) // 2 + 1

        month = month + 1

    return HebrewDate(year, Months(month), day)
