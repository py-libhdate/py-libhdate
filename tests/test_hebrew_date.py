"""Tests for the HebrewDate class."""

import datetime as dt

import pytest
from hypothesis import given, strategies

from hdate.hebrew_date import HebrewDate
from hdate.htables import Months

MIN_YEAR = 3762
MAX_YEAR = 6000


@strategies.composite
def valid_hebrew_date(draw: strategies.DrawFn) -> HebrewDate:
    """Generate a valid Hebrew date."""
    year = draw(strategies.integers(min_value=MIN_YEAR, max_value=MAX_YEAR))

    months = list(Months)
    if HebrewDate(year).is_leap_year():
        months.remove(Months.ADAR)
    else:
        months.remove(Months.ADAR_I)
        months.remove(Months.ADAR_II)
    month = draw(strategies.sampled_from(months))

    days = HebrewDate(year).days_in_month(month)
    day = draw(strategies.integers(min_value=1, max_value=days))

    return HebrewDate(year, month, day)


@strategies.composite
def relative_hebrew_date(draw: strategies.DrawFn) -> HebrewDate:
    """Generate a Hebrew date with no year."""
    month = draw(strategies.sampled_from(list(Months)))
    max_day = HebrewDate().days_in_month(month)
    day = draw(strategies.integers(min_value=1, max_value=max_day))
    return HebrewDate(0, month, day)


@given(d1=valid_hebrew_date(), d2=valid_hebrew_date())
def test_hebrew_date_comparisons(d1: HebrewDate, d2: HebrewDate) -> None:
    """Test that the HebrewDate class implements all comparison operators correctly."""
    assert (d1 == d2) == ((d1.year, d1.month, d1.day) == (d2.year, d2.month, d2.day))
    assert (d1 != d2) == ((d1.year, d1.month, d1.day) != (d2.year, d2.month, d2.day))
    assert (d1 < d2) == ((d1.year, d1.month, d1.day) < (d2.year, d2.month, d2.day))
    assert (d1 > d2) == ((d1.year, d1.month, d1.day) > (d2.year, d2.month, d2.day))
    assert (d1 <= d2) == ((d1.year, d1.month, d1.day) <= (d2.year, d2.month, d2.day))
    assert (d1 >= d2) == ((d1.year, d1.month, d1.day) >= (d2.year, d2.month, d2.day))


@given(d1=valid_hebrew_date(), d2=relative_hebrew_date())
def test_relative_hebrew_date_comparisons(d1: HebrewDate, d2: HebrewDate) -> None:
    """Test HebrewDatecomparison operators when there is no year."""
    assert (d1 == d2) == ((d1.month, d1.day) == (d2.month, d2.day))
    assert (d1 != d2) == ((d1.month, d1.day) != (d2.month, d2.day))
    assert (d1 < d2) == ((d1.month, d1.day) < (d2.month, d2.day))
    assert (d1 > d2) == ((d1.month, d1.day) > (d2.month, d2.day))
    assert (d1 <= d2) == ((d1.month, d1.day) <= (d2.month, d2.day))
    assert (d1 >= d2) == ((d1.month, d1.day) >= (d2.month, d2.day))


@pytest.mark.xfail(reason="Needs support for leap years")
@given(
    d1=valid_hebrew_date(),
    days=strategies.integers(min_value=-500, max_value=500),
)
def test_hebrew_date_addition(d1: HebrewDate, days: int) -> None:
    """Test HebrewDate addition and subtraction."""
    delta = dt.timedelta(days=days)
    d2 = d1 + delta
    assert d2 - d1 == delta


@pytest.mark.xfail(reason="Needs support for leap years")
@given(d1=valid_hebrew_date(), d2=relative_hebrew_date())
def test_hebrew_date_addition_with_no_year(d1: HebrewDate, d2: HebrewDate) -> None:
    """Test HebrewDate addition and subtraction when there is no year."""
    diff = d2 - d1
    assert d1 + diff == d2
