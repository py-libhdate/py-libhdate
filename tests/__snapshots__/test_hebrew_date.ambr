# serializer version: 1
# name: test_from_gdate_snapshot
  HebrewDate(year=5785, month=<Months.TEVET: 4>, day=5)
# ---
# name: test_from_jdn_snapshot
  HebrewDate(year=5760, month=<Months.TEVET: 4>, day=23)
# ---
# name: test_to_gdate_snapshot
  datetime.date(2025, 1, 5)
# ---
# name: test_to_jdn_snapshot
  2460681
# ---
